sunpy/net/tests/test_attr.py

```diff
@@ -1,27 +1,210 @@
+        a1.collides(1)
+
+
+def test_attr_or():
+    a1 = SA1(1)
+    a2 = SA2(2)
+
+    an = a1 | a2
+
+    assert isinstance(an, attr.AttrOr)
+    assert a1 in an.attrs
+    assert a2 in an.attrs
+    assert len(an.attrs) == 2
+
+    a1 = SA1(1)
+    a2 = SA2(1)
+
+    an = a1 | a2
+
+    assert an is a1
+
+
+def test_simpleattr_collides():
+    a1 = SA1(1)
+
+    with pytest.raises(TypeError):
+        a1 & a1
+
+
+def test_simple_attr_repr():
+    a1 = SA1("test string")
+
+    assert "test string" in repr(a1)
+    assert "SA1" in repr(a1)
+
 
 def test_dummyattr():
     one = attr.DummyAttr()
     other = attr.ValueAttr({'a': 'b'})
     assert (one | other) is other
     assert (one & other) is other
 
+
+def test_dummyattr_hash():
+    one = attr.DummyAttr()
+    assert hash(one) == hash(None)
+
+
+def test_dummyattr_collides():
+    one = attr.DummyAttr()
+    two = attr.DummyAttr()
+    assert one.collides(two) is False
+
+
+def test_dummyattr_eq():
+    one = attr.DummyAttr()
+    two = attr.DummyAttr()
+    other = attr.ValueAttr({'a': 'b'})
+    assert one == two
+    assert one != other
+
+
 def test_and_nesting():
-    a = attr.and_(attrs.Level(0),
-                  attr.AttrAnd((attrs.Instrument('EVE'),
-                                attrs.Time("2012/1/1", "2012/01/02"))))
+    a1 = SA1(1)
+    a2 = SA2(2)
+    a3 = SA3(3)
+
+    a = attr.and_(a1, attr.AttrAnd((a2, a3)))
     # Test that the nesting has been removed.
     assert len(a.attrs) == 3
 
+
 def test_or_nesting():
-    a = attr.or_(attrs.Instrument('a'),
-                  attr.AttrOr((attrs.Instrument('b'),
-                               attrs.Instrument('c'))))
+    a1 = SA1(1)
+    a2 = SA2(2)
+    a3 = SA3(3)
+
+    a = attr.or_(a1, attr.AttrOr((a2, a3)))
     # Test that the nesting has been removed.
     assert len(a.attrs) == 3
+
+
+def test_attr_metamagic(AIA, HMI, SPEC):
+    attr.Attr.update_values({Instrument: [('AIA', 'This is AIA, it takes data')]})
+    # .name is the attribute name return
+    assert attr.Attr._attr_registry[Instrument].name == [AIA.value.lower()]
+    # .name_long is the original name
+    assert attr.Attr._attr_registry[Instrument].name_long == [AIA.value]
+    # .des is the description of the item.
+    assert attr.Attr._attr_registry[Instrument].desc == ['This is AIA, it takes data']
+    # The _value_registry on the Attr object does not get cleaned.
+    # So by adding it again to the same type, in this case Instrument the list is appended.
+    attr.Attr.update_values({Instrument: [('HMI', 'This is HMI, it lives next to AIA')]})
+    assert attr.Attr._attr_registry[Instrument].name == [AIA.value.lower(), HMI.value.lower()]
+    assert attr.Attr._attr_registry[Instrument].name_long == [AIA.value, HMI.value]
+    assert attr.Attr._attr_registry[Instrument].desc == ['This is AIA, it takes data', 'This is HMI, it lives next to AIA']
+
+    # # Tests the print out for the first two inputs only
```

Comment:
I would split this up into sections.

Also I would define a fixture for `Attr`

```python
def EmptyAttr():
    empty = copy.copy(Attr._attr_registry)
    yield Attr
    Attr._attr_registry = empty
```

to make sure that each test get's a clean registry to test with.